surveyor/observation.go

```diff
@@ -161,90 +179,551 @@ func (o *serviceObsOptions) Validate() error {
 	return errors.New(strings.Join(errs, ", "))
 }
 
-// NewServiceObservation creates a new performance observation listener
-func NewServiceObservation(f string, sopts Options, metrics *ServiceObsMetrics, reconnectCtr *prometheus.CounterVec) (*ServiceObsListener, error) {
+// NewServiceObservationFromFile creates a new performance observation listener.
+func NewServiceObservationFromFile(f string, sopts Options, metrics *ServiceObsMetrics, reconnectCtr *prometheus.CounterVec) (*ServiceObsListener, error) {
 	js, err := os.ReadFile(f)
 	if err != nil {
 		return nil, err
 	}
 
-	opts := &serviceObsOptions{}
+	opts := &ObservationConfig{}
 	err = json.Unmarshal(js, opts)
 	if err != nil {
 		return nil, fmt.Errorf("invalid service observation configuration: %s: %s", f, err)
 	}
-
 	err = opts.Validate()
 	if err != nil {
 		return nil, fmt.Errorf("invalid service observation configuration: %s: %s", f, err)
 	}
 
-	sopts.Name = fmt.Sprintf("%s (observing %s)", sopts.Name, opts.ServiceName)
-	sopts.Credentials = opts.Credentials
-	sopts.Nkey = opts.Nkey
+	serviceObservation := &ServiceObservation{
+		ID:                f,
+		ObservationConfig: *opts,
+	}
+	obs, err := newServiceObservation(*serviceObservation, sopts, metrics, reconnectCtr)
+	if err != nil {
+		return nil, err
+	}
+
+	return obs, nil
+}
+
+func newServiceObservation(serviceObservation ServiceObservation, sopts Options, metrics *ServiceObsMetrics, reconnectCtr *prometheus.CounterVec) (*ServiceObsListener, error) {
+	err := serviceObservation.Validate()
+	if err != nil {
+		return nil, fmt.Errorf("invalid service observation configuration: %s: %s", serviceObservation.ServiceName, err)
+	}
+
+	sopts.Name = fmt.Sprintf("%s (observing %s)", sopts.Name, serviceObservation.ServiceName)
+	sopts.Credentials = serviceObservation.Credentials
+	sopts.Nkey = serviceObservation.Nkey
 	nc, err := connect(&sopts, reconnectCtr)
 	if err != nil {
 		return nil, fmt.Errorf("nats connection failed: %s", err)
 	}
 
 	return &ServiceObsListener{
-		nc:      nc,
-		logger:  sopts.Logger,
-		opts:    opts,
-		metrics: metrics,
-		sopts:   &sopts,
+		nc:          nc,
+		logger:      sopts.Logger,
+		observation: &serviceObservation,
+		metrics:     metrics,
+		sopts:       &sopts,
 	}, nil
 }
 
-// Start starts listening for observations
+func (s *Surveyor) startObservationsInDir() fs.WalkDirFunc {
+	return func(path string, info fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if filepath.Ext(info.Name()) != ".json" {
+			return nil
+		}
+
+		obs, err := NewServiceObservationFromFile(path, s.opts, s.observationMetrics, s.reconnectCtr)
+		if err != nil {
+			return fmt.Errorf("could not create observation from %s: %s", path, err)
+		}
+
+		// Prevent an equal observation to be loaded twice
+		// This is a problem that occurs with k8s mounts
+		for _, existingObservation := range s.observations {
+			if obs.observation.ServiceName == existingObservation.observation.ServiceName {
```

Comment:
I do not know if this is the best idea...  now that we are running a FS Watcher, we could have race conditions with this.  For example:

```
a/obs.json
```

Then you execute `mv a b`, if the events fire in this order:

1. Create dir `b` and start observations -> `b/obs.json` has an equal servicename as `a/obs.json`, skip
2. Delete dir `a` -> delete service from `a/obs.json`

Now you have a file with `b/obs.json` but no observation because of the ordering.

The original issue is #39 and we could work around that problem by ignoring hidden directories and files with the leading `.`

Reply:
I agree, that should be enough.